module/state_synchronization/indexer/in_memory_indexer.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	start := time.Now()
+
+	events := make([]flow.Event, 0)
+	results := make([]flow.LightTransactionResult, 0)
+	registers := make(map[ledger.Path]*ledger.Payload)
+	indexedCollections := 0
+
+	// Process all chunk data in a single pass
+	for idx, chunk := range data.ChunkExecutionDatas {
+		// Collect events
+		events = append(events, chunk.Events...)
+
+		// Collect transaction results
+		results = append(results, chunk.TransactionResults...)
+
+		// Process collections (except system chunk)
+		if idx < len(data.ChunkExecutionDatas)-1 {
+			if err := i.indexCollection(chunk.Collection); err != nil {
+				return fmt.Errorf("could not handle collection: %w", err)
+			}
+			indexedCollections++
+		}
+
+		// Process register updates
+		if chunk.TrieUpdate != nil {
+			// Verify trie update integrity
+			if len(chunk.TrieUpdate.Paths) != len(chunk.TrieUpdate.Payloads) {
+				return fmt.Errorf("update paths length is %d and registers length is %d and they don't match",
+					len(chunk.TrieUpdate.Paths), len(chunk.TrieUpdate.Payloads))
+			}
+
+			// Collect registers (last one for a path wins)
+			for i, path := range chunk.TrieUpdate.Paths {
+				registers[path] = chunk.TrieUpdate.Payloads[i]
+			}
+		}
+	}
+
+	if err := i.events.Store(data.BlockID, []flow.EventsList{events}); err != nil {
+		return fmt.Errorf("could not index events: %w", err)
+	}
+
+	if err := i.results.Store(data.BlockID, results); err != nil {
+		return fmt.Errorf("could not index transaction results: %w", err)
+	}
+
+	if err := i.indexRegisters(registers, i.header.Height); err != nil {
+		return fmt.Errorf("could not index registers: %w", err)
+	}
+
+	duration := time.Since(start)
+
+	i.metrics.BlockIndexed(
+		i.header.Height,
+		duration,
+		len(events),
+		len(registers),
+		len(results),
+	)
+
+	log.Debug().
+		Dur("duration_ms", duration).
+		Int("event_count", len(events)).
+		Int("register_count", len(registers)).
+		Int("result_count", len(results)).
+		Int("collection_count", indexedCollections).
+		Msg("indexed block data")
+
+	return nil
+}
+
+// indexRegisters processes register payloads and stores them in the register database.
+func (i *InMemoryIndexer) indexRegisters(registers map[ledger.Path]*ledger.Payload, height uint64) error {
+	regEntries := make(flow.RegisterEntries, 0, len(registers))
+
+	for _, register := range registers {
+		k, err := register.Key()
+		if err != nil {
+			return err
+		}
+
+		id, err := convert.LedgerKeyToRegisterID(k)
+		if err != nil {
+			return err
+		}
+
+		regEntries = append(regEntries, flow.RegisterEntry{
+			Key:   id,
+			Value: register.Value(),
+		})
+	}
+
+	return i.registers.Store(regEntries, height)
+}
+
+// indexCollection processes a collection and its associated transactions.
+func (i *InMemoryIndexer) indexCollection(collection *flow.Collection) error {
+	// Store the full collection
+	if err := i.collections.Store(collection); err != nil {
+		return err
```

Comment:
please wrap plain errors with enough context to debug the issue later.
```suggestion
		return fmt.Errorf("failed to store collection: %w", err)
```